### DISCUSSION TITLE
New rule for checking that a str is not assigned the variable None (at initialization)

### DISCUSSION THREAD

**Discussion by nunokaeru** [astral-sh/ruff#4456] [Ideas]
So I, myself, have the habit of doing `foo: str = None` which has been advised against.

I don't plan on adding more work to the mantainers, so I will implement the Rule myself, the Ruff Rule I propose is initially

```python

@dataclass
class Foo:
    bar: str = None
             ^ Ruff Error: Initializing str as None  
```

and it should allow auto-fixing to:

```python

@dataclass
class Foo:
    bar: str = ""
```

This could be improved upon with new rules later but let's keep it simple.
If I get the approval of the maintainers I shall start working on this,

**Comment by charliermarsh** [comment]
I appreciate your willingness to implement this yourself, that is awesome :)

The challenge with an autofix here is that the "correct" fix is ambiguous. E.g., I think the "correct" fix would more often (but not always) be `bar: str | None = None`, since `bar: str = None` _used_ to be acceptable to type checkers [prior to PEP 484](https://github.com/python/peps/pull/689), which removed that "implicit optional" behavior. So, I could see us adding a rule against implicit optionals like this, but I don't know that we'd be able to reliably fix it. We _could_ make a suggestion, but I'd vote that the suggestion to be to make optional explicit (`bar: str | None = None` in line with PEP 484) rather than change the base type (which also doesn't generalize behind primitives like `str`).

**Comment by nunokaeru** [comment]
I looked through https://peps.python.org/pep-0484/ and it looks like it was implemented in python 3.5
```
Python-Version: 3.5
```

Ruff itself only supports 3.7 and above and according to google 3.7 was released in `Python 3.7 was released on June 27, 2018`. 
However the [commit](https://github.com/python/peps/pull/689) that amends PEP484 was made in [July](https://github.com/python/peps/pull/689#event-1724595030).

So I assume that people using Ruff, would be using a version of 3.7 that is compatible with the proposed autofix. I assume for compatibility sakes in the code it would still be better if the autofix made the optional behaviour of the variable explicit and then it would be up to the person to change it from an optional themselves.

Will try working on this soon. Just wanted to give more context to the change.

**Comment by nunokaeru** [comment]
It looks like this rule as been implemented since the time we talked about it as the `implicit optional` rule

ref: https://github.com/astral-sh/ruff/blob/608df9a1bc0e6025049add877d1d833f1739e966/crates/ruff_linter/src/rules/ruff/rules/implicit_optional.rs

Thanks for your hard work, will close down this issue (and maybe look for some other rule to work on)
